src/Control/Retry.hs

```diff
@@ -246,29 +301,30 @@ capDelay limit p = RetryPolicyM $ \ n ->
 --
 -- Note how the latest failing result is returned after all retries
 -- have been exhausted.
-retrying :: MonadIO m
-         => RetryPolicyM m
-         -> (Int -> b -> m Bool)
-         -- ^ An action to check whether the result should be retried.
-         -- If True, we delay and retry the operation.
-         -> m b
-         -- ^ Action to run
-         -> m b
-retrying (RetryPolicyM policy) chk f = go 0
-    where
-      go n = do
-          res <- f
-          chk' <- chk n res
-          case chk' of
-            True -> do
-              chk <- policy n
-              case chk of
-                Just delay -> do
-                  liftIO (threadDelay delay)
-                  go $! n+1
-                Nothing -> return res
-            False -> return res
-
+retrying  :: MonadIO m
+          => RetryPolicyM m
+          -> (RetryStatus -> b -> m Bool)
+          -- ^ An action to check whether the result should be retried.
+          -- If True, we delay and retry the operation.
+          -> (RetryStatus -> m b)
+          -- ^ Action to run
+          -> m b
+retrying (RetryPolicyM policy) chk f = go defaultRetryStatus
+  where
+    go s = do
+        res <- f s
+        chk' <- chk s res
+        case chk' of
+          True -> do
+            chk <- policy s
+            case chk of
+              Just delay -> do
+                liftIO (threadDelay delay)
+                go $! RetryStatus { rsIterNumber = rsIterNumber s + 1
+                                  , rsCumulativeDelay = rsCumulativeDelay s + delay
+                                  , rsPreviousDelay = Just (maybe 0 (const delay) (rsPreviousDelay s))}
```

Comment:
Just ran a little test with 100K lines with and without the strict annotations and the difference was clear:

With strictness annotation, 9MB memory in use
```
cabal exec -- ghc --make -rtsopts -prof -fprof-auto Main.hs && ./Main +RTS -sstderr
     132,045,960 bytes allocated in the heap
      23,428,040 bytes copied during GC
       4,272,224 bytes maximum residency (5 sample(s))
         192,416 bytes maximum slop
               9 MB total memory in use (0 MB lost due to fragmentation)
```

Without strictness annotation, 38MB in use:
```
 cabal exec -- ghc --make -rtsopts -prof -fprof-auto Main.hs && ./Main +RTS -sstderr
[1 of 1] Compiling Main             ( Main.hs, Main.o ) [Control.Retry changed]
Linking Main ...
     136,845,960 bytes allocated in the heap
      77,852,048 bytes copied during GC
      18,744,504 bytes maximum residency (7 sample(s))
         536,560 bytes maximum slop
              38 MB total memory in use (0 MB lost due to fragmentation)
```